planner/core/rule_partition_processor.go

```diff
@@ -37,28 +38,35 @@ import (
 type partitionProcessor struct{}
 
 func (s *partitionProcessor) optimize(lp LogicalPlan) (LogicalPlan, error) {
-	// NOTE: partitionProcessor will assume all filter conditions are pushed down to
-	// DataSource, there will not be a Selection->DataSource case, so the rewrite just
-	// handle the DataSource node.
-	return s.rewriteDataSource(lp)
+	return s.rewriteDataSource(nil, lp)
 }
 
-func (s *partitionProcessor) rewriteDataSource(lp LogicalPlan) (LogicalPlan, error) {
+func (s *partitionProcessor) rewriteDataSource(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
 	// Assert there will not be sel -> sel in the ast.
 	switch lp.(type) {
 	case *DataSource:
-		return s.prune(lp.(*DataSource))
+		return s.prune(sel, lp.(*DataSource))
+	case *LogicalSelection:
+		return s.rewriteDataSource(lp.(*LogicalSelection), lp.Children()[0])
 	default:
 		children := lp.Children()
 		for i, child := range children {
-			newChild, err := s.rewriteDataSource(child)
+			newChild, err := s.rewriteDataSource(nil, child)
 			if err != nil {
 				return nil, errors.Trace(err)
 			}
 			children[i] = newChild
 		}
 	}
 
+	return s.selectOnSomething(sel, lp)
+}
+
+func (s *partitionProcessor) selectOnSomething(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
```

Comment:
This function would always return nil for error, shall we remove it from return field?